File: src/cogs/chal/chal.py
import typing

import interactions
from interactions import Extension, SlashContext

from lib.util import subcommand, get_ctf_forum
from lib.config import FORUM_GENERAL_CHANNEL


class Chal(Extension):
    '''Commands for managing challenges in a CTF forum'''

    @subcommand(name={"description": "The name of the challenge"}, category={"description": "the category of the challenge", "autocomplete": True})
    async def create(self, ctx: SlashContext, name: str, category: str) -> None:
        '''Creates a channel for the challenge'''
        forum = await get_ctf_forum(ctx)
        if (forum == None or ctx.channel.name != FORUM_GENERAL_CHANNEL):
            await ctx.send(":x: Must be used inside a CTF forum's general channel.")
            return
        if ("unsolved" in category.lower()):
            await ctx.send(":x: Unsolved cannot be a category.")
            return
        for post in forum.get_posts(exclude_archived=False):
            post_name = post.name or ""
            # remove solved prefix
            if (post_name.startswith("✔-")):
                post_name = post_name[len("✔-"):]
            if (post_name.lower() == name.lower()):
                await ctx.send(f":x: Challenge {post.mention} already exists.")
                return

        # check valid category
        if (not any(tag.name == category for tag in forum.available_tags)):
            await ctx.send(f":x: Could not find category {category}.")
            return

        post = await forum.create_post(name=name, content=name, applied_tags=[category, "unsolved", "unsolved-"+category])
        await ctx.send(f"{ctx.user.mention} created {post.mention}.")

    @create.autocomplete("category")
    async def get_categories(self, ctx: interactions.AutocompleteContext):
        '''Autocomplete that provides challenge categories for chal create'''
        forum = await get_ctf_forum(ctx)
        if (forum == None or ctx.channel.name != FORUM_GENERAL_CHANNEL):
            await ctx.send(["Must be used inside a CTF forum's general channel."])
            return

        categories = [tag.name for tag in forum.available_tags if (ctx.input_text in tag.name) and ("unsolved" not in tag.name)]
        await ctx.send(categories)

    @subcommand(flag={"description": "The flag for the challenge"})
    async def solve(self, ctx: SlashContext, flag: str) -> None:
        '''Marks a challenge as solved with a flag'''
        forum = await get_ctf_forum(ctx)
        if (forum == None):
            await ctx.send(":x: Must be used inside a CTF forum.")
            return
        if (ctx.channel.name == FORUM_GENERAL_CHANNEL):
            await ctx.send(":x: Cannot be used inside a CTF forum's general channel.")
            return
        post = typing.cast(interactions.GuildForumPost, ctx.channel)
        if (not any(tag.name == "unsolved" for tag in post.applied_tags)):
            await ctx.send(":x: Challenge does not have unsolved tag.")
            return

        tags = [tag for tag in post.applied_tags if "unsolved" not in tag.name]
        await ctx.send("Marking challenge as solved.")
        await post.edit(name="✔-" + (post.name or ""), applied_tags=tags, archived=True)

        # send message in general
        for post in forum.get_posts(exclude_archived=False):
            if (post.name == FORUM_GENERAL_CHANNEL):
                await post.send(f"{ctx.user.mention} solved challenge {ctx.channel.mention} with ||{flag}||")
                return

File: server/src/commands/chal.ts
import {
  ApplicationCommandOptionType,
  ApplicationCommandType,
} from "discord-api-types/v10";
import {
  createChannelMessage,
  createForumPost,
  editChannel,
  listForumPosts,
} from "../discord/client.js";
import { autocomplete } from "../discord/responses.js";
import { getCtfContext } from "./ctf-context.js";
import { deferredTask, finish } from "./helpers.js";
import { focusedOption, stringOption, subcommand, userId } from "./options.js";
import type { CommandDefinition } from "./types.js";

const GENERAL = "General";

export const chalCommand: CommandDefinition = {
  command: {
    name: "chal",
    type: ApplicationCommandType.ChatInput,
    description: "No Description Set",
    options: [
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "create",
        description: "Creates a channel for the challenge",
        options: [
          {
            type: ApplicationCommandOptionType.String,
            name: "name",
            description: "The name of the challenge",
            required: true,
          },
          {
            type: ApplicationCommandOptionType.String,
            name: "category",
            description: "the category of the challenge",
            required: true,
            autocomplete: true,
          },
        ],
      },
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "solve",
        description: "Marks a challenge as solved with a flag",
        options: [
          {
            type: ApplicationCommandOptionType.String,
            name: "flag",
            description: "The flag for the challenge",
            required: true,
          },
        ],
      },
    ],
  },
  async handle(interaction, runtime) {
    const action = subcommand(interaction)?.name;
    return deferredTask(interaction, runtime, false, async () => {
      const context = await getCtfContext(runtime.env, interaction, true);
      if (!context) {
        await finish(
          interaction,
          runtime,
          ":x: Must be used inside a CTF forum.",
        );
        return;
      }

      if (action === "create") {
        if (context.post.name !== GENERAL) {
          await finish(
            interaction,
            runtime,
            ":x: Must be used inside a CTF forum's general channel.",
          );
          return;
        }
        const category = stringOption(interaction, "category");
        if (category.toLowerCase().includes("unsolved")) {
          await finish(
            interaction,
            runtime,
            ":x: Unsolved cannot be a category.",
          );
          return;
        }
        const name = stringOption(interaction, "name");
        const posts = await listForumPosts(
          runtime.env,
          interaction.guild_id!,
          context.forum.id,
        );
        const duplicate = posts.find((post) => {
          const postName = post.name?.startsWith("✔-")
            ? post.name.slice(2)
            : post.name;
          return postName?.toLowerCase() === name.toLowerCase();
        });
        if (duplicate) {
          await finish(
            interaction,
            runtime,
            `:x: Challenge <#${duplicate.id}> already exists.`,
          );
          return;
        }
        const tags = context.forum.available_tags ?? [];
        const categoryTag = tags.find(
          (tag) => tag.name.toLowerCase() === category.toLowerCase(),
        );
        if (!categoryTag) {
          await finish(
            interaction,
            runtime,
            `:x: Could not find category ${category}.`,
          );
          return;
        }
        const unsolved = tags.find((tag) => tag.name === "unsolved");
        const unsolvedCategory = tags.find(
          (tag) =>
            tag.name.toLowerCase() ===
            `unsolved-${categoryTag?.name ?? category}`.toLowerCase(),
        );
        const post = await createForumPost(
          runtime.env,
          context.forum.id,
          name,
          name,
          [categoryTag, unsolved, unsolvedCategory]
            .filter((tag) => Boolean(tag))
            .map((tag) => tag!.id),
        );
        await finish(
          interaction,
          runtime,
          `<@${userId(interaction)}> created <#${post.id}>.`,
        );
        return;
      }

      if (context.post.name === GENERAL) {
        await finish(
          interaction,
          runtime,
          ":x: Cannot be used inside a CTF forum's general channel.",
        );
        return;
      }
      const tags = context.forum.available_tags ?? [];
      const appliedTags = (context.post.applied_tags ?? [])
        .map((id) => tags.find((tag) => tag.id === id))
        .filter(Boolean);
      if (!appliedTags.some((tag) => tag!.name.includes("unsolved"))) {
        await finish(
          interaction,
          runtime,
          `:x: Challenge does not have unsolved tag. Has tags: ${appliedTags.map((tag) => tag!.name).join(", ")}`,
        );
        return;
      }

      const remainingTagIds = appliedTags
        .filter((tag) => !tag!.name.includes("unsolved"))
        .map((tag) => tag!.id);
      await finish(interaction, runtime, "Marking challenge as solved.");
      await editChannel(runtime.env, context.post.id, {
        name: `✔-${context.post.name ?? ""}`,
        applied_tags: remainingTagIds,
        archived: true,
      });
      const posts = await listForumPosts(
        runtime.env,
        interaction.guild_id!,
        context.forum.id,
      );
      const general = posts.find((post) => post.name === GENERAL);
      if (general) {
        await createChannelMessage(
          runtime.env,
          general.id,
          `<@${userId(interaction)}> solved challenge <#${context.post.id}> with ||${stringOption(interaction, "flag")}||`,
        );
      }
    });
  },
  async autocomplete(interaction, runtime) {
    const context = await getCtfContext(runtime.env, interaction);
    if (!context || context.post.name !== GENERAL) {
      return autocomplete([
        {
          name: "Must be used inside a CTF forum's general channel.",
          value: "Must be used inside a CTF forum's general channel.",
        },
      ]);
    }
    const current = String(focusedOption(interaction)?.value ?? "");
    return autocomplete(
      (context.forum.available_tags ?? [])
        .filter(
          (tag) =>
            tag.name.toLowerCase().includes(current.toLowerCase()) &&
            !tag.name.toLowerCase().includes("unsolved"),
        )
        .map((tag) => ({ name: tag.name, value: tag.name })),
    );
  },
};
